### DISCUSSION TITLE
Nested Type Constraints

### DISCUSSION THREAD

**Discussion by Hvass-Labs** [pydantic/pydantic#4110] [Feature Request]
I am new to pydantic and maybe I have misunderstood how it works, but it seems that we cannot nest type-constraints, which would be a very useful feature.

For example, we can do the following to ensure that a list of ints has constraints on both the list and the ints:

    from pydantic import BaseModel, conlist, confloat

    class Foo(BaseModel):
        bar : conlist(confloat(gt=123.45), min_items=2, max_items=3)
        
    Foo(bar=[234.5, '356.7', 456.7])
    # Result: Foo(bar=[234.5, 356.7, 456.7])

    Foo(bar=[234.5, '356.7', 456.7, 567.8])
    # ValidationError: Max 3 items in list.
    
    Foo(bar=[34.5, '356.7', 456.7])
    # ValidationError: Value must be greater than 123.45

In a previous post #4109 I requested a new data-type called `FloatScaled` which could turn a string like '678.9k' into the float 678_900 because the letter 'k' would be interpreted as kilo (i.e. a factor 1000). It would be very useful if we could further constrain this value using `confloat`, so we could do something like this:

    class Bing(BaseModel):
        bong : confloat(FloatScaled, gt=123_000)

As I understand, the only way of doing that currently, would be to make a new function e.g. `confloat_scaled` which would merge the functionality of `confloat` and `FloatScaled`. But this would have to be done for every combination of data-type and constraint that users might be interested in, so a more modular approach would be desirable.

Thanks!

**Comment by samuelcolvin** [comment]
I think this is what annotated types should be good at once that is released and integrated with pydantic - https://github.com/annotated-types/annotated-types

**Comment by Hvass-Labs** [comment]
Thanks! That looks exactly like what I need. I don't want to put pressure on you, but do you have an estimated time-line for when this might be included in Pydantic? Are we talking maybe 1 month, or possibly 6 months? Just so I know whether I should wait a while for this to be featured in Pydantic, or build my own work-around in the meantime.

**Reply by samuelcolvin** [reply]
It'll hopefully be in V2, which will take >1 month, <6.
